# Exploratory data analysis of the tech survey results data set

In [1]:
import sys
import time

import matplotlib.pyplot as plt

%matplotlib inline
import numpy as np
import pandas as pd
from IPython.display import HTML

sys.path.append("code/.")

import mglearn
from IPython.display import display
# from plotting_functions import *

# Classifiers and regressors
from sklearn.dummy import DummyClassifier, DummyRegressor

# Preprocessing and pipeline
from sklearn.impute import SimpleImputer

# train test split and cross validation
from sklearn.model_selection import cross_val_score, cross_validate, train_test_split
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import (
    MinMaxScaler,
    OneHotEncoder,
    OrdinalEncoder,
    StandardScaler,
)
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier


In [2]:
tech_df = pd.read_csv('../data/raw/survey_results_ca_usa.csv')

In [3]:
tech_df.shape

(11600, 48)

In [4]:
tech_df.shape[0] - tech_df.dropna().shape[0]

11600

# Summary of the data set
The data set used in this project is from the salary and work information survey created by Stack Overflow in 2019. Each row in the data set represents work related information of a software developer, including the job title, education level, location, years of experience and technology, etc. All the data was collected from the online survey in 2019. There are 11600 observations in the data set, and 48 features. All the observations have at least one missing values in the data set. Below we show the number of missing values for each feature in the data set.

In [5]:
pd.DataFrame(tech_df.isnull().sum()).rename(columns = {0:'missing_value_counts'})

missing_value_counts
ResponseId                                       0
MainBranch                                       0
Employment                                       4
Country                                          0
US_State                                      1808
UK_Country                                   11600
EdLevel                                          4
Age1stCode                                       6
LearnCode                                       25
YearsCode                                       55
YearsCodePro                                    72
DevType                                         24
OrgSize                                         18
Currency                                         0
CompTotal                                        0
CompFreq                                         0
LanguageHaveWorkedWith                          22
LanguageWantToWorkWith                         757
DatabaseHaveWorkedWith                        1625
DatabaseWantToWorkWith                        3309
PlatformHaveWorkedWith                        2971
PlatformWantToWorkWith                        4619
WebframeHaveWorkedWith                        2643
WebframeWantToWorkWith                        4079
MiscTechHaveWorkedWith                        5116
MiscTechWantToWorkWith                        6382
ToolsTechHaveWorkedWith                        941
ToolsTechWantToWorkWith                       1938
NEWCollabToolsHaveWorkedWith                   128
NEWCollabToolsWantToWorkWith                  1182
OpSys                                            5
NEWStuck                                        18
NEWSOSites                                      13
SOVisitFreq                                     34
SOAccount                                       24
SOPartFreq                                    2130
SOComm                                          34
NEWOtherComms                                   36
Age                                             44
Gender                                          65
Trans                                          108
Sexuality                                      673
Ethnicity                                      173
Accessibility                                  438
MentalHealth                                   430
SurveyLength                                   152
SurveyEase                                      96
ConvertedCompYearly                              0

# Partition the data set into training and test sets
Before proceeding further, we will split the data such that 80% of observations are in the training and 20% of observations are in the test set. Below we list the summary and head as well as tail for the training data set:

In [6]:
pd.set_option("max_columns", 50)
train_df, test_df = train_test_split(tech_df, test_size=0.2, random_state=123)

In [7]:
train_df.describe(include = 'all')

ResponseId                      MainBranch          Employment  \
count    9280.000000                            9280                9276   
unique           NaN                               2                   5   
top              NaN  I am a developer by profession  Employed full-time   
freq             NaN                            8402                8623   
mean    41178.040302                             NaN                 NaN   
std     24510.640982                             NaN                 NaN   
min        18.000000                             NaN                 NaN   
25%     19769.500000                             NaN                 NaN   
50%     44456.500000                             NaN                 NaN   
75%     60547.250000                             NaN                 NaN   
max     83438.000000                             NaN                 NaN   

                         Country    US_State  UK_Country  \
count                       9280        7859         0.0   
unique                         2          52         NaN   
top     United States of America  California         NaN   
freq                        7924        1039         NaN   
mean                         NaN         NaN         NaN   
std                          NaN         NaN         NaN   
min                          NaN         NaN         NaN   
25%                          NaN         NaN         NaN   
50%                          NaN         NaN         NaN   
75%                          NaN         NaN         NaN   
max                          NaN         NaN         NaN   

                                             EdLevel     Age1stCode LearnCode  \
count                                           9278           9276      9263   
unique                                             9              9       343   
top     Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years    School   
freq                                            5257           5179      1025   
mean                                             NaN            NaN       NaN   
std                                              NaN            NaN       NaN   
min                                              NaN            NaN       NaN   
25%                                              NaN            NaN       NaN   
50%                                              NaN            NaN       NaN   
75%                                              NaN            NaN       NaN   
max                                              NaN            NaN       NaN   

       YearsCode YearsCodePro                DevType  \
count       9237         9221                   9263   
unique        52           52                   2096   
top           10            5  Developer, full-stack   
freq         712          708                   1440   
mean         NaN          NaN                    NaN   
std          NaN          NaN                    NaN   
min          NaN          NaN                    NaN   
25%          NaN          NaN                    NaN   
50%          NaN          NaN                    NaN   
75%          NaN          NaN                    NaN   
max          NaN          NaN                    NaN   

                         OrgSize                   Currency     CompTotal  \
count                       9267                       9280  9.280000e+03   
unique                        10                         11           NaN   
top     10,000 or more employees  USD\tUnited States dollar           NaN   
freq                        1893                       7952           NaN   
mean                         NaN                        NaN  1.512036e+05   
std                          NaN                        NaN  1.165809e+06   
min                          NaN                        NaN  1.000000e+00   
25%                          NaN                        NaN  8.200000e+04   
50%                          NaN     

In [8]:
train_df.head()

ResponseId                      MainBranch          Employment  \
7966        57095  I am a developer by profession  Employed full-time   
10813       77069  I am a developer by profession  Employed full-time   
9452        66495  I am a developer by profession  Employed full-time   
10338       73864  I am a developer by profession  Employed full-time   
2331        12841  I am a developer by profession  Employed full-time   

                        Country    US_State  UK_Country  \
7966   United States of America  California         NaN   
10813                    Canada         NaN         NaN   
9452   United States of America  Washington         NaN   
10338  United States of America    Nebraska         NaN   
2331   United States of America  California         NaN   

                                                 EdLevel     Age1stCode  \
7966        Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
10813  Some college/university study without earning ...  18 - 24 years   
9452     Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  11 - 17 years   
10338       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
2331        Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   

                                               LearnCode YearsCode  \
7966          School;Online Forum;Books / Physical media        24   
10813  Coding Bootcamp;Online Courses or Certificatio...         6   
9452                       School;Books / Physical media        17   
10338                      School;Books / Physical media        23   
2331                                              School         7   

      YearsCodePro                                            DevType  \
7966            15                              Developer, full-stack   
10813            4          Developer, full-stack;Developer, back-end   
9452             9  Developer, full-stack;Engineer, data;Data scie...   
10338           20         Developer, front-end;Developer, full-stack   
2331             2                              Developer, full-stack   

                        OrgSize                   Currency  CompTotal  \
7966   1,000 to 4,999 employees  USD\tUnited States dollar     500000   
10813      500 to 999 employees       CAD\tCanadian dollar     100000   
9452         20 to 99 employees  USD\tUnited States dollar     275000   
10338  1,000 to 4,999 employees  USD\tUnited States dollar     150000   
2331   10,000 or more employees  USD\tUnited States dollar      83000   

      CompFreq                             LanguageHaveWorkedWith  \
7966    Yearly  Bash/Shell;Java;JavaScript;Kotlin;Node.js;Obje...   
10813   Yearly                            JavaScript;Node.js;Ruby   
9452    Yearly  Bash/Shell;Clojure;Java;JavaScript;Node.js;Sca...   
10338   Yearly    HTML/CSS;Java;JavaScript;Node.js;SQL;TypeScript   
2331    Yearly       C#;HTML/CSS;JavaScript;Python;SQL;TypeScript   

                                  LanguageWantToWorkWith  \
7966   JavaScript;Kotlin;Node.js;Python;SQL;Swift;Typ...   
10813                     Clojure;Elixir;Go;Haskell;Rust   
9452                                    Clojure;Rust;SQL   
10338    HTML/CSS;Java;JavaScript;Node.js;SQL;TypeScript   
2331     C#;Go;HTML/CSS;JavaScript;Python;SQL;TypeScript   

                  DatabaseHaveWorkedWith      DatabaseWantToWorkWith  \
7966                  MySQL;Redis;SQLite       Firebase;Redis;SQLite   
10813      MySQL;PostgreSQL;Redis;SQLite                         NaN   
9452   DynamoDB;Elasticsearch;PostgreSQL                  PostgreSQL   
10338                              MySQL               MongoDB;MySQL   
2331          Microsoft SQL Server;Redis  Microsoft SQL Server;Redis   

                       PlatformHaveWorkedWith     PlatformWantToWorkWith  \
7966                AWS;Google Cloud Platform      Google Cloud Platform   
10813  AWS;DigitalOcean;Google Cloud Platform                        NaN   
9452                      

In [9]:
train_df.tail()

ResponseId                                         MainBranch  \
9785       69393                     I am a developer by profession   
7763       56274                     I am a developer by profession   
5218       36475                     I am a developer by profession   
1346        7260  I am not primarily a developer, but I write co...   
3582       25835  I am not primarily a developer, but I write co...   

              Employment                   Country        US_State  \
9785  Employed full-time  United States of America           Texas   
7763  Employed full-time  United States of America      California   
5218  Employed full-time  United States of America      Washington   
1346  Employed full-time  United States of America        Colorado   
3582  Employed full-time  United States of America  South Carolina   

      UK_Country                                            EdLevel  \
9785         NaN       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
7763         NaN    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
5218         NaN         Other doctoral degree (Ph.D., Ed.D., etc.)   
1346         NaN       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3582         NaN  Some college/university study without earning ...   

         Age1stCode                                          LearnCode  \
9785  18 - 24 years  Coding Bootcamp;Friend or family member;Books ...   
7763  11 - 17 years                                             School   
5218  18 - 24 years  Other online resources (ex: videos, blogs, etc...   
1346  11 - 17 years  Other online resources (ex: videos, blogs, etc...   
3582  18 - 24 years  School;Friend or family member;Books / Physica...   

     YearsCode      YearsCodePro  \
9785         1  Less than 1 year   
7763        26                21   
5218        44                46   
1346        12                12   
3582        21                16   

                                                DevType  \
9785                              Developer, full-stack   
7763  Developer, full-stack;Product manager;Senior E...   
5218  Developer, front-end;Developer, desktop or ent...   
1346  Engineer, data;Data scientist or machine learn...   
3582  Developer, desktop or enterprise applications;...   

                       OrgSize                   Currency  CompTotal CompFreq  \
9785      100 to 499 employees  USD\tUnited States dollar      70000   Yearly   
7763        20 to 99 employees  USD\tUnited States dollar     325000   Yearly   
5218  1,000 to 4,999 employees  USD\tUnited States dollar     125000  Monthly   
1346  1,000 to 4,999 employees  USD\tUnited States dollar     100000   Yearly   
3582  10,000 or more employees  USD\tUnited States dollar     127460   Weekly   

                        LanguageHaveWorkedWith  \
9785  C++;Dart;HTML/CSS;JavaScript;Node.js;SQL   
7763              HTML/CSS;JavaScript;Ruby;SQL   
5218                     C#;Haskell;TypeScript   
1346                   Bash/Shell;Python;R;SQL   
3582         C#;HTML/CSS;JavaScript;Python;VBA   

                    LanguageWantToWorkWith  \
9785                          Clojure;Java   
7763        Elixir;Node.js;Rust;TypeScript   
5218  C#;Erlang;Haskell;Node.js;TypeScript   
1346               Bash/Shell;Python;R;SQL   
3582                 C#;Node.js;TypeScript   

                                 DatabaseHaveWorkedWith  \
9785                       Firebase;MongoDB;MySQL;Redis   
7763                     Elasticsearch;PostgreSQL;Redis   
5218                               Microsoft SQL Server   
1346                    MariaDB;MySQL;PostgreSQL;SQLite   
3582  Cassandra;Microsoft SQL Server;MongoDB;MySQL;O...   

                                 DatabaseWantToWorkWith  \
9785                                                NaN   
7763                                           Firebase   
5218                 Microsoft SQL Server;MongoDB;Redis   
1346                                         PostgreSQL   

In [10]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9280 entries, 7966 to 3582
Data columns (total 48 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ResponseId                    9280 non-null   int64  
 1   MainBranch                    9280 non-null   object 
 2   Employment                    9276 non-null   object 
 3   Country                       9280 non-null   object 
 4   US_State                      7859 non-null   object 
 5   UK_Country                    0 non-null      float64
 6   EdLevel                       9278 non-null   object 
 7   Age1stCode                    9276 non-null   object 
 8   LearnCode                     9263 non-null   object 
 9   YearsCode                     9237 non-null   object 
 10  YearsCodePro                  9221 non-null   object 
 11  DevType                       9263 non-null   object 
 12  OrgSize                       9267 non-null   object 
 13  

# Counts of observations for each feature
Since we only want to look into the top five significant features of salary prediction, it is good to know the counts distribution for each feature. The tables below show the counts of different role which is quite imbalanced.

In [11]:
pd.DataFrame(train_df["MainBranch"].value_counts()).rename(columns = {'MainBranch':'counts'})

counts
I am a developer by profession                        8402
I am not primarily a developer, but I write cod...     878

There are 7924 observations from United States of America and 1356 observations from Canada which is also imbalanced here.

In [12]:
pd.DataFrame(train_df["Country"].value_counts()).rename(columns = {'Country':'counts'})

counts
United States of America    7924
Canada                      1356

In education level feature, we can see most developers have hold a bachelor's degree.

In [13]:
pd.DataFrame(train_df["EdLevel"].value_counts()).rename(columns = {'EdLevel':'counts'})

counts
Bachelor’s degree (B.A., B.S., B.Eng., etc.)          5257
Master’s degree (M.A., M.S., M.Eng., MBA, etc.)       1669
Some college/university study without earning a...    1187
Associate degree (A.A., A.S., etc.)                    439
Other doctoral degree (Ph.D., Ed.D., etc.)             372
Secondary school (e.g. American high school, Ge...     197
Something else                                          74
Professional degree (JD, MD, etc.)                      50
Primary/elementary school                               33

Experience is important when considering the salary of a developer, the table below shows the counts of developer having different years of coding experience.

In [14]:
yearcode_df = pd.DataFrame(train_df["YearsCode"].value_counts()).reset_index()#.rename(columns={"index":"code_years", "YearsCode":"count"}, inplace=True)
yearcode_df = yearcode_df.rename(columns={"index":"code_years", "YearsCode":"count"})
yearcode_df = (
                yearcode_df[(yearcode_df['code_years'] != "Less than 1 year") & (yearcode_df['code_years'] != "More than 50 years")]
               .apply(pd.to_numeric).sort_values(by = 'code_years').set_index('code_years')
              )#.set_index('code_years').apply(pd.to_numeric).sort_index(inplace=True)
yearcode_df

count
code_years       
1              27
2              88
3             147
4             242
5             423
6             478
7             469
8             489
9             338
10            712
11            307
12            402
13            267
14            231
15            500
16            241
17            211
18            229
19            109
20            540
21            180
22            181
23            174
24            127
25            371
26            101
27            102
28             81
29             34
30            266
31             44
32             69
33             64
34             49
35            203
36             50
37             65
38             80
39             56
40            187
41             56
42             48
43             33
44             26
45             46
46             22
47             12
48              7
49              8
50             12

Coding language might also have effect on the target salary, the table shows the counts of developer using different programming languages. We can see Javascript is the most frequently used language for developers.

In [15]:
language_df = pd.DataFrame(train_df["LanguageHaveWorkedWith"].str.split(';',expand=True).apply(pd.value_counts).sum(axis=1))
language_df 

0
APL            34.0
Assembly      376.0
Bash/Shell   3610.0
C            1360.0
C#           2992.0
C++          1705.0
COBOL          47.0
Clojure       284.0
Crystal        67.0
Dart          224.0
Delphi        103.0
Elixir        291.0
Erlang         96.0
F#            127.0
Go           1257.0
Groovy        430.0
HTML/CSS     5186.0
Haskell       153.0
Java         2621.0
JavaScript   6297.0
Julia         129.0
Kotlin        604.0
LISP          153.0
Matlab        272.0
Node.js      3330.0
Objective-C   298.0
PHP          1321.0
Perl          346.0
PowerShell   1487.0
Python       4460.0
R             443.0
Ruby         1160.0
Rust          841.0
SQL          4970.0
Scala         265.0
Swift         504.0
TypeScript   3326.0
VBA           493.0